---
title: "R_RNA"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE, fig.path='figures/', fig.width=8, fig.height=5 )
```
by adding `fig.path = 'figures/'` we put all of the figures created when we knit this document into a directory called `figures`


# Differential Expression Testing

Read the docs: https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

Installs:
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager") # calls the package from the source
BiocManager::install("GSEAbase")
BiocManager::install("clusterProfiler")
install.packages("devtools")
install.packages("RColorBrewer")
install.packages("pheatmap")
devtools::install_github("karthik/wesanderson")
BiocManager::install("org.Sc.sgd.db")
BiocManager::install("GOstats")
BiocManager::install("edgeR")
BiocManager::install("tximport")
install.packages("treemap")
install.packages("tidyverse")
install.packages("ggplot2")
```


Load Libraries: 
```{r, warning = FALSE, message = FALSE}
library(tximport)
library(DESeq2)
library(tidyverse)
library("GSEABase")
library(clusterProfiler)
library(RColorBrewer)
library(pheatmap)
library(wesanderson)
library(org.Sc.sgd.db)
library(GOstats)
library(edgeR)
library(treemap)
library(tidyverse)
library(ggplot2)
```

Import sample metadata: 
```{r, warning = FALSE, message = FALSE, cache=TRUE}
# read in the file from url
samples <- read_csv("https://osf.io/cxp2w/download")
# look at the first 6 lines
samples
```

```{r}
samples <- samples %>% mutate(quant_file = str_sub(quant_file, 3))

samples
```

Import tx 2 gene file: 
```{r}
#tx2gene_map <- read_tsv("https://osf.io/a75zm/download")

tx <- read.table("quant/ERR458493.qc.fq.gz_quant/quant.sf", header = TRUE)
tx2gene_map <- data.frame(tx$Name, tx$Name)
names(tx2gene_map) <-  c("TXNAME", "GENEID")

txi <- tximport(files = samples$quant_file, type = "salmon", tx2gene = tx2gene_map)
colnames(txi$counts) <- samples$sample
```

Make DESeq2 object: 
```{r}
dds <- DESeqDataSetFromTximport(txi = txi, 
                                colData = samples, 
                                design = ~condition)
dds$condition <- relevel(dds$condition, ref = "wt") # make wild-type the reference to which expression in treatment samples is compared to 
```

Run DESeq2: 
```{r, cache = TRUE}
dds <- DESeq(dds)
```

Check out results: 
```{r}
res <- results(dds)
head(res)
```

Summarize results
```{r}
summary(res, alpha = 0.05) # default significance cut-off is 0.1, changing alpha to 0.05 changes the significance cut-off 
```

**Why p-adj?** Multiple testing problems come up when doing multiple statistical tests simultaneously. Here, we are performing 9,474 individual statistical tests. We first need to think about the definifition p-value: the p-value is the probability of attaining the observed result by chance. So if we use a p-value cut off of 0.05, 5% of the time our results will represent chance outcomes rather than real effects. This is fine in one test. For that one test we are 95% confident that are results are real. But if we do 9,474 tests, 473 genes could be *significantly* differentially expressed, just by chance -- there is a false positive problem.

FDR (False Discovery Rate) adjusted p-values controls for the expected rate of false positives. The p-adj is almost always higher than the p-value. Controlling for multiple testing has the downside of increasing the possibility of false negatives. 

Some researchers also set a cut-off for minimum expression and minimum log fold-changes.

If you want to use a log fold-change cut off:
```{r}
res_lfcut <- results(dds, lfcThreshold = 1, altHypothesis = "greaterAbs")
summary(res_lfcut, alpha = 0.05)
```


Lets say you wanted the comparison to be in the other direction:
```{r}
res_rev <- results(dds, contrast=c("condition","wt", "snf2"))
head(res_rev)
```


# Visualizing RNA-seq results 

## Normalization

**Count Data Transformations:** 
for ranking and visualizations (e.g. PCA plots and heatmaps)

**rlog**: "transforms the count data to the log2 scale in a way which minimizes differences between samples for rows with small counts, and which normalizes with respect to library size. The rlog transformation produces a similar variance stabilizing effect as varianceStabilizingTransformation, though rlog is more robust in the case when the size factors vary widely. The transformation is useful when checking for outliers or as input for machine learning techniques such as clustering or linear discriminant analysis." -- from function documentation 

This is computationally very time intensive. 

```{r, cache=TRUE}
rld <- rlog(dds, blind=TRUE)
head(assay(rld), 3)
```

** Variance stabilizing transformation (so much faster than rlog):**
"This function calculates a variance stabilizing transformation (VST) from the fitted dispersion-mean relation(s) and then transforms the count data (normalized by division by the size factors or normalization factors), yielding a matrix of values which are now approximately homoskedastic (having constant variance along the range of mean values). The transformation also normalizes with respect to library size. The rlog is less sensitive to size factors, which can be an issue when size factors vary widely. These transformations are useful when checking for outliers or as input for machine learning techniques such as clustering or linear discriminant analysis."" – from function documentation

```{r, cache = TRUE}
vsd <- vst(dds, blind = TRUE)
head(assay(vsd), 3)
```

## Ordination

rlog PCA: 
```{r pca_rld}
data1 <- plotPCA(rld, returnData=TRUE)
data1$group<-gsub(" : ","_",as.character(data1$group))
percentVar1 <- round(100 * attr(data1, "percentVar"))

PCA<-ggplot(data1, aes(PC1, PC2, color = condition))+ theme_bw()+
  geom_point(size=9, alpha = 0.8) + scale_colour_manual(values = c("#44aabb","#bbbbbb"))+
  xlab(paste0("PC1: ",percentVar1[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar1[2],"% variance")) +
  theme(text = element_text(size=20)) + ggtitle("rlog PCA")
PCA
#ggsave("figures/vsd_PCA.png", device="png") # to save the plot
```

variance stabilized PCA:
```{r pca_vst}
data1 <- plotPCA(vsd, returnData=TRUE)
data1$group<-gsub(" : ","_",as.character(data1$group))
percentVar1 <- round(100 * attr(data1, "percentVar"))

PCA<-ggplot(data1, aes(PC1, PC2, color = condition))+ theme_bw()+
  geom_point(size=9, alpha = 0.8) + scale_colour_manual(values = c("#44aabb","#bbbbbb"))+
  xlab(paste0("PC1: ",percentVar1[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar1[2],"% variance")) +
  theme(text = element_text(size=20)) + ggtitle("vst PCA")
PCA
#ggsave("figures/vsd_PCA.png", device="png") # to save the plot
```

## HeatMaps

rlog HeatMap:
```{r heatmap_rld}

df <- as.data.frame(colData(rld)[,c("condition", "sample")])

mat_colors1<-list(sample = brewer.pal(12, "Paired")[0:6])
names(mat_colors1$sample)<- df$sample

mat_colors <- list(condition = brewer.pal(12, "Paired")[7:8])
names(mat_colors$condition) <- c("wt", "snf2")

genes <- order(res$padj)[1:1000]

 pheatmap(assay(rld)[genes, ], cluster_rows=TRUE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df, annotation_colors = c(mat_colors1, mat_colors), fontsize = 12)
```

variance stabilized HeatMap: 
```{r heatmap_vst}
df <- as.data.frame(colData(vsd)[,c("condition", "sample")])

pheatmap(assay(vsd)[genes, ], cluster_rows=TRUE, show_rownames=FALSE, show_colnames = FALSE,
         cluster_cols=FALSE, annotation_col=df, annotation_colors = c(mat_colors1, mat_colors), fontsize = 12)
```

Another option for heat maps: 
plot the difference from the mean normalized count across samples 
(and optionally change default colors)

With Rlog transformed data:
```{r heatmap_rld_meandiff}
library(wesanderson)
pal <- wes_palette(name = "Zissou1", n=2000 , type= "continuous")

mat_colors1<-list(sample = wes_palette("IsleofDogs1", 6))
names(mat_colors1$sample)<- df$sample

mat_colors <- list(condition = wes_palette("Cavalcanti1")[4:5])
names(mat_colors$condition) <- c("wt", "snf2")

mat <- assay(rld)[genes, ]
mat <- mat - rowMeans(mat)

df <- as.data.frame(colData(rld)[,c("condition", "sample")])

pheatmap(mat,  cluster_rows=TRUE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col=df, annotation_colors = c(mat_colors1, mat_colors), fontsize = 12, color = pal)

```

Same but with variance stabilizing function:
```{r heatmap_vst_meandiff}
mat <- assay(vsd)[genes, ]
mat <- mat - rowMeans(mat)

df <- as.data.frame(colData(vsd)[,c("condition", "sample")])

pheatmap(mat,  cluster_rows=TRUE, show_rownames=FALSE, show_colnames = FALSE,
         cluster_cols=FALSE, annotation_col=df, annotation_colors = c(mat_colors1, mat_colors), fontsize = 12, color = pal)

```


Heatmap of sample-to-sample distances
```{r heatmap_sampledistance}
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(vsd$condition, vsd$type, sep="-")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors)
```


## MA plot

```{r}
DESeq2::plotMA(res, alpha = 0.05, ylim = c(-2,2))
#default alpha is 0.1
```

```{r}
DESeq2::plotMA(vsd)
```


# Gene Set Enrichment Testing 
If you remember, we had  598 significantly upregulated genes and 1049 significantly down regulated genes in this data set (this is pretty typical). That is a lot to try to make sense of. If you know you are interested in a specific gene or a specific pathway, you can look for that in your data, but if you are trying to figure out what is generally different betwene treatments, it helps to categaorize and summarize genes by what they do. Two common ways to do this are GO terms and KEGG pathways.

```{r}
summary(res, alpha = 0.05)
```

## GO term enrichment

"A GO annotation is a statement about the function of a particular gene. GO annotations are created by associating a gene or gene product with a GO term. Together, these statements comprise a “snapshot” of current biological knowledge. Hence, GO annotations capture statements about how a gene functions at the molecular level, where in the cell it functions, and what biological processes (pathways, programs) it helps to carry out.

Different pieces of knowledge regarding gene function may be established to different degrees, which is why each GO annotation always refers to the evidence upon which it is based. All GO annotations are ultimately supported by the scientific literature, either directly or indirectly. In GO, the supporting evidence is presented in the form of a GO Evidence Codes and either a published reference or description of the methodology used to create the annotation. The GO evidence codes describe the type of evidence and reflect how far removed the annotated assertion is from direct experimental evidence, and whether this evidence was reviewed by an expert biocurator."  -- http://geneontology.org/docs/go-annotations/


```{r}
GO_df = toTable(org.Sc.sgdGO)
head(GO_df)
```

This frame comes with all three types of GO terms in one frame, BP = Biological Process, MF = Molecular Function, CC = Cellular COmponenet  


Convert the df to the format required for GOstats:
```{r}
goframeData = data.frame(GO_df$go_id, GO_df$Evidence, GO_df$systematic_name)
names(goframeData) = c("GO", "Evidence", "gene_id")
goframeData$GO <- as.character(goframeData$GO)
goframeData$Evidence <- as.character(goframeData$Evidence)
goframeData$gene_id <- as.character((goframeData$gene_id))
head(goframeData)
```

Now turn this into a GO frame for GOstats:
```{r, warning = FALSE, message= FALSE}
goFrame<- GOFrame(goframeData)
goAllFrame<- GOAllFrame(goFrame)
gsc <- GeneSetCollection(goAllFrame, setType = GOCollection())
```

### GO terms enriched in UP regulated genes:

get data frame of DESeq2 results:
```{r}
DEres <- as.data.frame(res)
```

```{r}
upFrame <- DEres[DEres$padj <= 0.05 & DEres$log2FoldChange >= 1,]
upFrame <- upFrame[complete.cases(upFrame),]
dim(upFrame)
```
(by selecting genes with a log2 fold change >1 or <-1, we decrease the number of genes we consider --> this is a 2x change )
we now have 175 upregulated genes to consider (rather than 598)

```{r}
universe = Lkeys(org.Sc.sgdGO)

genes = row.names(upFrame)
params <- GSEAGOHyperGParams(name="Yeast, wt v. snf2", geneSetCollection= gsc, geneIds = genes, universeGeneIds = universe, ontology = "BP", pvalueCutoff = 0.05, conditional = TRUE, testDirection = "over")

Over <- hyperGTest(params)

Over
```

## Enrichr (edgeR package)

```{r}
SC <- toTable(org.Sc.sgdGENENAME)

DEres <- as.data.frame(res)
DEres$systematic_name<- row.names(DEres)
DEres<- merge(DEres, SC, by="systematic_name")

gene <- DEres$gene_name

gene.df <- bitr(gene, fromType = "GENENAME",
        toType = c("ENTREZID", "ENSEMBL"),
        OrgDb = org.Sc.sgd.db)
names(gene.df)<- c("gene_name", "EntrezID", "ENSEMBL" )

go.df <- bitr(gene, fromType = "GENENAME",
        toType = c("GO"),
        OrgDb = org.Sc.sgd.db)
names(go.df)<- c("gene_name", "GO", "evidence", "ontology")

DEres<- merge(DEres, gene.df, by = "gene_name" )

go_dict <- merge(go.df, gene.df, by= "gene_name")

BPdict <- go_dict[go_dict$ontology == "BP", ]

BPdict<- BPdict[,c(2,6)]

upFrame <- DEres[DEres$padj <= 0.05 & DEres$log2FoldChange > 0.0,]
upFrame <- upFrame[complete.cases(upFrame),]

universe<- DEres$systematic_name
gene <- upFrame$systematic_name

enrchdUP_GO<-enricher(gene, pAdjustMethod = "none", universe, pvalueCutoff = 0.05,TERM2GENE= BPdict)

upBP_GOs <- data.frame(enrchdUP_GO)

upBP_GOs
```

```{r}
upBP_GOs$ID
```

*could do the same the thing with Molecular Function (MF) or cellular component (CC) GO terms*

### REVIGO
http://revigo.irb.hr/

REVIGO is an online tool that helps summarize and visualize long lists of GO terms 

```{r}
revigo.names <- c("term_ID","description","freqInDbPercent","uniqueness","dispensability","representative");
revigo.data <- rbind(c("GO:0016192","vesicle-mediated transport",1.085,0.567,0.000,"vesicle-mediated transport"),
c("GO:0019344","cysteine biosynthetic process",0.131,0.502,0.000,"cysteine biosynthesis"),
c("GO:0006333","chromatin assembly or disassembly",0.120,0.236,0.311,"cysteine biosynthesis"),
c("GO:0042407","cristae formation",0.013,0.294,0.120,"cysteine biosynthesis"));

stuff <- data.frame(revigo.data);
names(stuff) <- revigo.names;

stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
stuff$freqInDbPercent <- as.numeric( as.character(stuff$freqInDbPercent) );
stuff$uniqueness <- as.numeric( as.character(stuff$uniqueness) );
stuff$dispensability <- as.numeric( as.character(stuff$dispensability) );

treemap(
	stuff,
	index = c("representative","description"),
	vSize = "uniqueness",
	type = "categorical",
	vColor = "representative",
	palette = wes_palette("BottleRocket2"),
	title = "REVIGO Gene Ontology treemap",
	inflate.labels = FALSE,      # set this to TRUE for space-filling group labels - good for posters
	lowerbound.cex.labels = 0,   # try to draw as many labels as possible (still, some small squares may not get a label)
	bg.labels = "#CCCCCCAA",     # define background color of group labels
												       # "#CCCCCC00" is fully transparent, "#CCCCCCAA" is semi-transparent grey, NA is opaque
	position.legend = "none"
)

```



## GUI Options 

Get upregulated gene IDs for David: https://david.ncifcrf.gov/gene2gene.jsp
```{r}
UpEnsemble <- upFrame$ENSEMBL[!is.na(upFrame$ENSEMBL)]
write.csv(UpEnsemble, "upEnsembleIDs.csv", row.names = FALSE, col.names = FALSE, quote = FALSE )
```
- Look in the files pane, open `upEnsembleIDs.csv`
- delete the "x" header
- copy all (command A)
- paste into DAVID website
**12 "clusters" of genes by function**


for GOrilla:
http://cbl-gorilla.cs.technion.ac.il/
        
- Look in the files pane, open `upEnsembleIDs.csv`          
- delete the "x" header           
- copy all (command A)            
- paste into GOrilla website        
- shows GO hierarchical organization        
- shows all enriched GO terms (nested)        
- function `enrichGO` in clusterProfiler package        

KEGG annotation and pathway enrichment is another type of functional enrichment that can be more useful than GO terms. 
https://www.kegg.jp/

some useful KEGG tools:
          
- function `kegga` in edgeR package (bioconductor)        
- pathview package (bioconductor)         
- iPath3, interactive: https://pathways.embl.de/        

